torch/_dynamo/metrics_context.py

```diff
@@ -0,0 +1,124 @@
+import time
+from contextlib import contextmanager
+from typing import Any, Callable, Dict, Generator, Optional, Type, Union
+from typing_extensions import TypeAlias
+
+
+OnExitType: TypeAlias = Callable[
+    [Dict[str, Any], Optional[Type[BaseException]], Optional[BaseException]], None
+]
+
+
+class MetricsContext:
+    def __init__(self, on_exit: OnExitType):
+        """
+        Use this class as a contextmanager to create a context under which to accumulate
+        a set of metrics, e.g., metrics gathered during a compilation. On exit of the
+        contextmanager, call the provided 'on_exit' function and pass a dictionary of
+        all metrics set during the lifetime of the contextmanager.
+        """
+        self._on_exit = on_exit
+        self._metrics: Dict[str, Any] = {}
+        self._start_time: Dict[str, int] = {}
+        self._recording = False
+
+    def __enter__(self) -> "MetricsContext":
+        """
+        Initialize metrics recording.
+        """
+        if self._recording:
```

Comment:
I have some failing tests on the PR above this one because of this check. So either I've put the context manager too low in that PR or we actually do need to support re-entering. If anyone has some insight...